packages/model-viewer/src/three-components/ARRenderer.ts

```diff
@@ -248,117 +327,330 @@ export class ARRenderer extends EventDispatcher {
     return this[$presentedScene] != null;
   }
 
-  async placeModel() {
-    if (this[$currentSession] == null) {
-      return;
+  [$updateCamera](view: XRView) {
+    const {camera} = this;
+    const {matrix: cameraMatrix} = camera;
+
+    cameraMatrix.fromArray(view.transform.matrix);
+    camera.updateMatrixWorld(true);
+    // position is not updated when matrix is updated.
+    camera.position.setFromMatrixPosition(cameraMatrix);
+
+    if (this[$initialHitSource] != null) {
+      // Target locked to screen center
+      const {position, model} = this[$presentedScene]!;
+      const radius = model.idealCameraDistance;
+      camera.getWorldDirection(position);
+      position.multiplyScalar(radius);
+      position.add(camera.position);
     }
-    // NOTE: Currently rays will be cast from the middle of the screen.
-    // Eventually we might use input coordinates for this.
 
-    // Just reuse the hit matrix that the reticle has computed.
-    if (this.reticle && this.reticle.hitMatrix) {
+    if (!this[$initialized]) {
+      camera.projectionMatrix.fromArray(view.projectionMatrix);
+      // Have to set the inverse manually when setting matrix directly. This is
+      // needed for raycasting.
+      camera.projectionMatrixInverse.getInverse(camera.projectionMatrix);
+      // Orient model toward camera on first frame.
+      const {x, z} = camera.position;
       const scene = this[$presentedScene]!;
-      const {model} = scene;
-      const {hitMatrix} = this.reticle;
+      scene.pointTowards(x, z);
+      scene.model.updateMatrixWorld(true);
+      this[$goalYaw] = scene.yaw;
+      this[$initialModel2World].copy(scene.model.matrixWorld);
+      this[$initialized] = true;
+    }
 
-      scene.position.setFromMatrixPosition(hitMatrix);
+    this[$presentedScene]!.model.orientHotspots(
+        Math.atan2(cameraMatrix.elements[1], cameraMatrix.elements[5]));
+  }
 
-      // Orient the scene to face the camera
-      const camPosition = vector3.setFromMatrixPosition(this.camera.matrix);
-      scene.pointTowards(camPosition.x, camPosition.z);
-      scene.updateMatrixWorld();
+  [$placeInitially](frame: XRFrame) {
+    const hitSource = this[$initialHitSource];
+    if (hitSource == null) {
+      return;
+    }
 
-      model.visible = true;
-      model.setHotspotsVisibility(true);
+    const hitTestResults = frame.getHitTestResults(hitSource);
+    if (hitTestResults.length == 0) {
+      return;
+    }
 
-      this.dispatchEvent({type: 'modelmove'});
+    const hit = hitTestResults[0];
+    const hitMatrix = this[$getHitPoint](hit);
+    if (hitMatrix == null) {
+      return;
     }
+
+    this.placeModel(hitMatrix);
+
+    hitSource.cancel();
+    this[$initialHitSource] = null;
+
+    const {session} = frame;
+    session.addEventListener('selectstart', this[$selectStartHandler]);
+    session.addEventListener('selectend', this[$selectEndHandler]);
```

Comment:
When do these listeners get removed?

Reply:
When we stop presenting we set the session to null and exit it, which I thought would mean it would get garbage-collected along with its listeners. Is that untrue?

Reply:
It's usually safer to remove them when you're done, which is to say when the `EventTarget` leaves your scope of influence. If you don't know how long an object will have a reference held to it, you risk leaking the whole `ARRenderer` and more.